guppylang/compiler/expr_compiler.py

```diff
@@ -452,68 +462,117 @@ def visit_ResultExpr(self, node: ResultExpr) -> Wire:
-            assert isinstance(gen.hasnext, PlaceNode)
-            inputs = [gen.iter, list_name]
-            with self._new_loop(inputs, gen.hasnext):
-                # If there is a next element, compile it and continue with the next
-                # generator
-                compiler.compile_stmts([gen.hasnext_assign], self.dfg)
-                with self._if_true(gen.hasnext, inputs):
-
-                    def compile_ifs(ifs: list[ast.expr]) -> None:
-                        """Helper function to compile a series of if-guards into nested
-                        Conditional nodes."""
-                        if ifs:
-                            if_expr, *ifs = ifs
-                            # If the condition is true, continue with the next one
-                            with self._if_true(if_expr, inputs):
-                                compile_ifs(ifs)
-                        else:
-                            # If there are no guards left, compile the next generator
-                            compile_generators(elt, gens)
-
-                    compiler.compile_stmts([gen.next_assign], self.dfg)
-                    compile_ifs(gen.ifs)
-
-            # After the loop is done, we have to finalize the iterator
-            self.visit(gen.iterend)
-
-        compile_generators(node.elt, node.generators)
-        return self.visit(list_name)
+        def build_update(elt_port: Wire) -> None:
+            """Callback for pushing a comprehension element onto the list."""
+            list_port = self.dfg[list_place]
+            (self.dfg[list_place],) = self._build_method_call(
+                list_ty, "append", node, [list_port, elt_port], list_ty.args
+            ).inout_returns
+
+        self._compile_generators(node.elt, node.generators, [list_place], build_update)
+        return self.dfg[list_place]
+
+    def visit_DesugaredArrayComp(self, node: DesugaredArrayComp) -> Wire:
+        # Allocate an uninitialised array of the desired size and a counter variable
+        array_ty = get_type(node)
+        assert isinstance(array_ty, OpaqueType)
+        array_var = Variable(next(tmp_vars), array_ty, node)
+        count_var = Variable(next(tmp_vars), int_type(), node)
+        hugr_elt_ty = (
+            ht.Option(node.elt_ty.to_hugr())
+            if node.elt_ty.linear
+            else node.elt_ty.to_hugr()
+        )
+        self.dfg[array_var] = self.builder.add_op(
+            array_new_uninitialized(hugr_elt_ty, node.length)
+        )
+        self.dfg[count_var] = self.builder.load(
+            hugr.std.int.IntVal(0, width=NumericType.INT_WIDTH)
+        )
+
+        def build_update(elt: Wire) -> None:
+            """Callback for putting an element into the array."""
+            array, count = self.dfg[array_var], self.dfg[count_var]
+            (self.dfg[array_var],) = self._build_method_call(
+                array_ty, "__setitem__", node, [array, count, elt], array_ty.args
+            ).inout_returns
```

Comment:
very-optional, probably for another PR - but how about `all_returns` that gives back a tuple of both inout and regular, so you can check the other is empty in your (non-exhaustive) patterm-match, and so you can't accidentally forget which one to call?
(I slightly dislike that which of `inout_returns` and `regular_returns` you want, depends on the *sugar*/style with which the function was declared, rather than the actual functionality, right?)

Reply:
It's a `NamedTuple`, so we can actually unpack it directly

Reply:
> I slightly dislike that which of inout_returns and regular_returns you want, depends on the sugar/style with which the function was declared, rather than the actual functionality, right?

Yes, but I don't see a way around that...